perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_adGroupName.java

```diff
@@ -25,6 +34,62 @@ public void checkAttributeSyntax(PerunSessionImpl sess, Group group, Attribute a
 		}
 	}
 
+	public void checkAttributeSemantics(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongAttributeAssignmentException, WrongReferenceAttributeValueException {
+		//Attribute can be null
+		if(attribute.getValue() == null) {
+			return;
+		}
+
+		List<Resource> assignedResources = sess.getPerunBl().getResourcesManagerBl().getAssignedResources(sess, group);
+
+		for (Resource assignedResource : assignedResources) {
+			try {
+				Attribute resourceAdRepresentation = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, assignedResource, A_R_D_AD_RESOURCE_REPRESENTATION);
+
+				if (resourceAdRepresentation.valueAsString().equals("tree")) {
+					if (!groupHasUniqueName(sess, group, attribute)) {
+						throw new WrongReferenceAttributeValueException(attribute, "There already exists a group with the same name and same hierarchy!");
+					}
+				}
+			} catch (AttributeNotExistsException e) {
+				//We can skip this case
+			}
+		}
+	}
+
+	/**
+	 * Checks uniqueness of adGroupName by comparing it to all adGroupNames of subroups of parent group.
+	 *
+	 * @param sess perun session
+	 * @param group group
+	 * @param attribute attribute
+	 * @return true if the name is unique, false otherwise
+	 * @throws AttributeNotExistsException
+	 * @throws WrongAttributeAssignmentException
+	 */
+	private boolean groupHasUniqueName(PerunSessionImpl sess, Group group, Attribute attribute) throws AttributeNotExistsException, WrongAttributeAssignmentException {
+		if (group.getParentGroupId() != null) {
+			try {
+				Group parentGroup = sess.getPerunBl().getGroupsManagerBl().getParentGroup(sess, group);
+
+				List<Group> subGroups = sess.getPerunBl().getGroupsManagerBl().getSubGroups(sess, parentGroup);
+				subGroups.remove(group);
+
+				for (Group subGroup : subGroups) {
+					Attribute nameAttribute = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, subGroup, A_G_D_AD_GROUP_NAME);
+					String name = nameAttribute.valueAsString();
+					if (name.equals(attribute.valueAsString())) {
```

Comment:
I am not sure if other groups must have this attribute set, therefore I would compare them like this to avoid `NullPointerException`:
```suggestion
					if (attribute.valueAsString().equals(name)) {
```

Reply:
Fixed.